src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,600 @@
+        cdef int count
+        cdef fmpq_mpoly res
+
+        if not PyDict_Check(d):
+            raise ValueError("expected a dictionary")
+        n = PyDict_Size(d)
+        exponents = <fmpz_struct *> libc.stdlib.calloc(nvars, sizeof(fmpz_struct))
+        if exponents == NULL:
+            raise MemoryError()
+        exp_ptr = <fmpz_struct **> libc.stdlib.calloc(nvars, sizeof(fmpz_struct *))
+        if exp_ptr == NULL:
+            libc.stdlib.free(exponents)
+            raise MemoryError()
+        for i in range(nvars):
+            fmpz_init(exponents + i)
+            exp_ptr[i] = exponents + i
+        # fmpq_init(coefficient)
+        res = fmpq_mpoly.__new__(fmpq_mpoly)
+        res.ctx = self
+        fmpq_mpoly_init(res.val, res.ctx.val)
+        res._init = True
+        count = 0
+        for k, v in d.items():
+            xtype = fmpq_set_any_ref(coefficient, v)
+            if xtype == FMPZ_UNKNOWN:
+                for i in range(nvars):
+                    fmpz_clear(exponents + i)
+                libc.stdlib.free(exponents)
+                libc.stdlib.free(exp_ptr)
+                raise TypeError("invalid coefficient type %s" % type(v))
+            if not PyTuple_Check(k):
+                for i in range(nvars):
+                    fmpz_clear(exponents + i)
+                libc.stdlib.free(exponents)
+                raise TypeError("Expected tuple of ints as key not %s" % type(k))
+            if PyTuple_GET_SIZE(k) != nvars:
+                for i in range(nvars):
+                    fmpz_clear(exponents + i)
+                libc.stdlib.free(exponents)
+                raise TypeError("Expected exponent tuple of length %d" % nvars)
+            for i, tup in enumerate(k):
+                xtype = fmpz_set_any_ref(exponents + i, tup)
+                if xtype == FMPZ_UNKNOWN:
+                    for i in range(nvars):
+                        fmpz_clear(exponents + i)
+                    libc.stdlib.free(exponents)
+                    raise TypeError("Invalid exponent type %s" % type(tup))
+            #Todo lobby for fmpz_mpoly_push_term_fmpz_ffmpz
+            if not fmpq_is_zero(coefficient):
+                fmpq_mpoly_push_term_fmpq_fmpz(res.val, coefficient, exp_ptr, self.val)
+                # _fmpq_mpoly_push_exp_ffmpz(res.val, exponents, self.val)
+                # fmpq_mpoly_set_term_coeff_fmpz(res.val, count, coefficient, self.val)
+                count += 1
+        for i in range(nvars):
+            fmpz_clear(exponents + i)
+            fmpq_mpoly_sort_terms(res.val, self.val)
+            fmpq_mpoly_reduce(res.val, self.val)
+        return res
+
+
+cdef class fmpq_mpoly(flint_mpoly):
+    """
+    The *fmpz_poly* type represents sparse multivariate polynomials over
+    the integers.
+    """
+
+    def __cinit__(self):
+        self._init = False
+
+    def __dealloc__(self):
+        if self._init:
+            fmpq_mpoly_clear(self.val, self.ctx.val)
+            self._init = False
```

Comment:
Is it necessary to set `_init = False` here? The object is being deallocated and does not exist after this method is called.

Reply:
Hmm, I think this is just being defensive, the Cython docs guarantee the `__cinit__` is called exactly once, however  the `__dealloc__` method doesn't have the same note. Though I think it is the case? It seems this method is called sometime after `__del__`, which isn't guaranteed by the Python spec to only be called once, however CPython docs say it is only called once, even if the object is resurrected. To make this even more unclear there's an open Cython bug report that may be caused by `__dealloc__` being called twice, though it seems much more edge-casey than our issue `https://github.com/cython/cython/issues/6022`.
https://docs.python.org/3/reference/datamodel.html#object.__del__
https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#finalization-method 

Actually it seems this has the potential to leak in the future, the `__dealloc__` method of a super class is always called as well, if any children of a class utilises `self._init` as well it may not deallocate the memory.
If this was just dealing with `malloc`/`free` I'd say we should just set the pointer to `NULL` and not to care if it's freed twice, but the flint clear functions absolutely do not no op on `NULL`.
Perhaps the variable should be mangled to avoid this, but then we'd need to bypass that mangling when using the `__new__` methods for faster initialisation.

Reply:
I don't think it is necessary to have `__dealloc__` in a superclass. We should just make sure that exactly one class in the inheritance hierarchy is the one that controls access to the C level data.